File: 2023/UICPC DIV1/UICPC Round #07/solutions/teamchange/submissions/accepted/teamchange-zf.py
#!/bin/python3

import sys

sys.setrecursionlimit(100000)

def add_edge(u, v, c):
    if v not in g[u]: g[u][v] = 0
    if u not in g[v]: g[v][u] = 0
    g[u][v] += c

def augment(v, pre, t, mn, par):
    if v == t: return mn
    if v in par: return 0
    par[v] = pre

    for w in g[v]:
        if g[v][w]:
            nxt = augment(w, v, t, min(mn, g[v][w]), par)
            if nxt:
                g[v][w] -= nxt
                g[w][v] += nxt
                return nxt
    return 0


N, R = map(int, input().split())
team = input()
change = input()

def In(v): return 2*v
def Out(v): return 2*v+1

g = {v : {} for v in range(2*N+2)}
s, t = 2*N, 2*N+1

for i in range(N):
    if change[i] != "?":
        if team[i] == change[i]: add_edge(s, In(i), 2)
        else: add_edge(Out(i), t, 2)
    add_edge(In(i), Out(i), 1)

for _ in range(R):
    u, v = map(int, input().split())
    u -= 1
    v -= 1
    add_edge(Out(u), In(v), 2)
    add_edge(Out(v), In(u), 2)

par = {}
while augment(s, -2, t, 10**9, par): par = {}

ans = {}
for i in range(N):
    if Out(i) in par: ans[i] = team[i]
    elif In(i) in par: ans[i] = 'X'
    else: ans[i] = {"A":"B", "B":"A"}[team[i]]
print("".join(ans[i] for i in range(N)))

File: 2023/UICPC DIV1/UICPC Round #07/solutions/teamchange/submissions/accepted/teamchange-zf.cpp
// Even Ford-Fukerson works: O(N^2 + N*R) for this particular problem since
// maxflow <= N and the graph has size O(N + R).

#include <bits/stdc++.h>

const int INF = 1e9; // set >= max capacity

using namespace std;

vector<unordered_map<int, int>> g;
vector<int> par;

void init(int n) { g = vector<unordered_map<int, int>>(n); }
void add_edge(int u, int v, int c) { g[u][v] = c; }

int augment(int v, int pre, int t, int mn) {
  if (v == t) return mn;
  if (par[v] != -1) return 0;
  par[v] = pre;

  for (auto& p : g[v]) {
    if (p.second == 0) continue;
    int nxt = augment(p.first, v, t, min(mn, p.second));
    if (nxt) {
      p.second -= nxt;
      g[p.first][v] += nxt;
      return nxt;
    }
  }
  return 0;
}

// augment until done, the mincut is stored in par, no need to know flow value
void mincut(int s, int t) {
  do { par.assign(g.size(), -1);
  } while(augment(s, -2, t, INF));
}

int main() {
  int N, R;
  cin >> N >> R;

  string team, change;
  cin >> team >> change;

  init(2*N+2);

  auto in = [](int i) { return 2*i; };
  auto out = [](int i) { return 2*i+1; };
  int s = 2*N, t = 2*N+1;

  for (int i = 0; i < N; ++i) {
    if (change[i] != '?') {
      if (team[i] == change[i]) add_edge(s, in(i), 2);
      else add_edge(out(i), t, 2);
    }
    add_edge(in(i), out(i), 1);
  }

  while (R--) {
    int u, v;
    cin >> u >> v;
    --u; --v;
    add_edge(out(u), in(v), 2);
    add_edge(out(v), in(u), 2);
  }

  mincut(s, t);

  string final(N, '_');
  for (int i = 0; i < N; ++i)
    if (par[out(i)] != -1) final[i] = team[i];
    else if (par[in(i)] != -1) final[i] = 'X';
    else final[i] = (team[i] == 'A' ? 'B' : 'A');
  cout << final << endl;

  return 0;
}
